overlord/snapstate/snapmgr.go

```diff
@@ -386,6 +418,96 @@ func (m *SnapManager) blockedTask(cand *state.Task, running []*state.Task) bool
 	return false
 }
 
+var CanAutoRefresh func(st *state.State) bool
+
+// ensureRefreshes ensures that we refresh all installed snaps periodically
+func (m *SnapManager) ensureRefreshes() error {
+	m.state.Lock()
+	defer m.state.Unlock()
+
+	// see if it even makes sense to try to refresh
+	if CanAutoRefresh == nil || !CanAutoRefresh(m.state) {
+		return nil
+	}
+
+	tr := config.NewTransaction(m.state)
+
+	// for tests
+	var refreshDisabled bool
+	err := tr.Get("core", "refresh.disabled", &refreshDisabled)
+	if err != nil && !config.IsNoOption(err) {
+		return err
+	}
+	if refreshDisabled {
+		return nil
+	}
+
+	var lastRefresh time.Time
+	err = tr.Get("core", "refresh.last", &lastRefresh)
+	if err != nil && !config.IsNoOption(err) {
+		return err
+	}
+
+	var refreshRand time.Duration
+	err = tr.Get("core", "refresh.rand", &refreshRand)
+	if err != nil && !config.IsNoOption(err) {
+		return err
+	}
+
+	nextRefresh := lastRefresh.Add(minRefreshInterval).Add(refreshRand)
+	if time.Now().Before(nextRefresh) {
+		return nil
+	}
+
+	// check that there is no change in flight already
+	for _, chg := range m.state.Changes() {
+		if chg.Kind() == "auto-refresh" && !chg.Status().Ready() {
+			// change already in motion
+			return nil
+		}
+	}
+
+	// For setLastRefresh() we have two options here:
+	//  1. setLastRefresh only if no error happened
+	//     below in the "AutoRefresh()" call
+	//  2. always set the last refresh regardless of errors
+	//     in "AutoRefresh()"
+	//
+	// This currently implements (2) because if there are e.g. store
+	// errors we don't want to put more burden on the store by having
+	// every client in the world retry every 10min.
+	//
+	// However (1) is better for e.g. network errors on the client
+	// side (not yet connected etc)
+	setLastRefresh(m.state)
+	updated, tasksets, err := AutoRefresh(m.state)
+	if err != nil {
+		return err
+	}
+
+	var msg string
+	switch len(updated) {
+	case 0:
+		// check in after some hours
+		logger.Noticef("No snaps to auto-refresh found")
+		return nil
+	case 1:
+		msg = fmt.Sprintf(i18n.G("Refresh snap %q"), updated[0])
+	default:
+		quoted := strutil.Quoted(updated)
+		// TRANSLATORS: the %s is a comma-separated list of quoted snap names
+		msg = fmt.Sprintf(i18n.G("Refresh snaps %s"), quoted)
```

Comment:
I changed it to `Refresh %d snaps`. I can add the first one or two names if you want but it seems a bit random.